server/src/main/java/org/elasticsearch/index/translog/TranslogStreamOutput.java

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.index.translog;
+
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.BytesRefIterator;
+import org.elasticsearch.common.io.stream.RecyclerBytesStreamOutput;
+import org.elasticsearch.common.recycler.Recycler;
+
+import java.io.IOException;
+
+/**
+ * This class provided specialized serialization methods for translog operations. The goal is to allow direct
+ * access to the current recycler bytes stream output page and write the entire operation with a single bounds
+ * check.
+ */
+public class TranslogStreamOutput extends RecyclerBytesStreamOutput {
+
+    public static final int FIXED_INDEX_HEADER_SIZE = 39;
+    public static final int FIXED_DELETE_HEADER_SIZE = 30;
+    public static final int FIXED_NO_OP_HEADER_SIZE = 21;
+
+    public TranslogStreamOutput(Recycler<BytesRef> recycler) {
+        super(recycler);
+    }
+
+    public void writeSerializedOperation(Translog.Serialized serialized) throws IOException {
```

Comment:
IIUC, this is only to avoid bounds checking, but from the looks of it we'd have only 3 such bounds checks anyway since serialized.iterator() is just 2 entries and then we have the checksum. Is this significant enough to go to this length? If so, I'd imagine we should remove the `refIterator` too to avoid the object construction?

Reply:
I mostly pulled this out as I don't think benchmarking justifies it at the moment.